File: examples/filter_hls/software/src/filter.cc
// Copyright 2018 Delft University of Technology
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <memory>
#include <vector>
#include <iostream>
#include <iomanip>
#include <random>
#include <stdlib.h>
#include <unistd.h>

// Apache Arrow
#include <arrow/api.h>

// Fletcher
#include "fletcher/api.h"

std::shared_ptr<arrow::Schema> getFilterReadSchema() {
  std::vector<std::shared_ptr<arrow::Field>> schema_fields = {
      arrow::field("first", arrow::utf8(), false),
      arrow::field("last", arrow::utf8(), false),
      arrow::field("zip", arrow::uint32(), false)
  };
  auto schema = std::make_shared<arrow::Schema>(schema_fields, metaMode(fletcher::Mode::READ));
  return schema;
}

std::shared_ptr<arrow::Schema> getFilterWriteSchema() {
  std::vector<std::shared_ptr<arrow::Field>> schema_fields = {
      arrow::field("first", arrow::utf8(), false),
  };
  auto schema = std::make_shared<arrow::Schema>(schema_fields, metaMode(fletcher::Mode::WRITE));
  return schema;
}

std::shared_ptr<arrow::RecordBatch> getFilterInputRB(int32_t num_entries) {
  // Some first names
  std::vector<std::string> first_names = {"Alice", "Bob", "Carol", "David"};
  std::vector<std::string> last_names = {"Cooper", "Smith", "Smith", "Smith"};
  std::vector<uint32_t> zip_codes = {1337, 4242, 1337, 1337};

  // Make a string builder
  arrow::StringBuilder fn_builder;
  arrow::StringBuilder ln_builder;
  arrow::UInt32Builder zip_builder;

  // Append the strings in the string builder
  if (!fn_builder.AppendValues(first_names).ok()) {
    throw std::runtime_error("Could not append to first name builder.");
  };
  if (!ln_builder.AppendValues(last_names).ok()) {
    throw std::runtime_error("Could not append to last name builder.");
  };
  if (!zip_builder.AppendValues(zip_codes).ok()) {
    throw std::runtime_error("Could not append zip codes to builder.");
  };

  // Array to hold Arrow formatted string data
  std::shared_ptr<arrow::Array> fn_array;
  std::shared_ptr<arrow::Array> ln_array;
  std::shared_ptr<arrow::Array> zip_array;

  // Finish building and create a new data array around the data
  if (!fn_builder.Finish(&fn_array).ok()) {
    throw std::runtime_error("Could not finalize first name builder.");
  };
  if (!ln_builder.Finish(&ln_array).ok()) {
    throw std::runtime_error("Could not finalize last name builder.");
  };
  if (!zip_builder.Finish(&zip_array).ok()) {
    throw std::runtime_error("Could not finalize zip code builder.");
  };

  // Create the Record Batch
  std::shared_ptr<arrow::RecordBatch> record_batch =
      arrow::RecordBatch::Make(getFilterReadSchema(), fn_array->length(), {fn_array, ln_array, zip_array});

  return record_batch;
}

std::shared_ptr<arrow::RecordBatch> getFilterOutputRB(int32_t num_entries, int32_t num_chars) {
  std::shared_ptr<arrow::Buffer> offsets;
  std::shared_ptr<arrow::Buffer> values;
  arrow::AllocateBuffer(num_entries + 1, &offsets);
  arrow::AllocateBuffer(num_chars, &values);
  auto sa = std::make_shared<arrow::StringArray>(2, offsets, values);
  std::shared_ptr<arrow::RecordBatch> record_batch = arrow::RecordBatch::Make(getFilterWriteSchema(), 2, {sa});
  return record_batch;
}

int main(int argc, char **argv) {

  constexpr int32_t num_entries = 4;

  auto rb_in = getFilterInputRB(num_entries);
  auto rb_out = getFilterOutputRB(4096, 4096);

  std::cout << "RecordBatch in: " << std::endl;
  for (int i = 0; i < rb_in->num_columns(); i++) {
    std::cout << rb_in->column(i)->ToString() << std::endl;
  }

  std::shared_ptr<fletcher::Platform> platform;
  std::shared_ptr<fletcher::Context> context;
  std::shared_ptr<fletcher::UserCore> uc;

  fletcher::Platform::Make(&platform);

  platform->init();

  fletcher::Context::Make(&context, platform);

  uc = std::make_shared<fletcher::UserCore>(context);

  context->queueRecordBatch(rb_in);
  context->queueRecordBatch(rb_out);

  uc->reset();

  context->enable();

  uc->setRange(0, num_entries);
  uc->setArguments({1337}); // zip code

  uc->start();

  uc->waitForFinish(10);

  // Get raw pointers to host-side Arrow buffers
  auto sa = std::dynamic_pointer_cast<arrow::StringArray>(rb_out->column(0));

  auto raw_offsets = sa->value_offsets()->mutable_data();
  auto raw_values = sa->value_data()->mutable_data();

  platform->copyDeviceToHost(context->device_arrays[5]->buffers[0].device_address, raw_offsets, 4096);
  platform->copyDeviceToHost(context->device_arrays[5]->buffers[1].device_address, raw_values, 4096);

  //fletcher::HexView hv(0);
  //hv.addData(raw_offsets, sizeof(int32_t) * (num_str+1));
  //hv.addData(raw_values, num_chars);
  //std::cout << hv.toString() << std::endl;
  std::cout << "RecordBatch out:" << std::endl;
  std::cout << sa->ToString() << std::endl;
}

File: examples/filter_hls/software/python/filter.py
# Copyright 2018 Delft University of Technology
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import gc
import timeit
import pandas as pd
import pyarrow as pa
import argparse
import numpy as np
import pyfletcher as pf

import filter_custom


class Timer:
    def __init__(self, gc_disable=True):
        self.starttime = 0
        self.stoptime = 0
        self.gc_disable = gc_disable

    def start(self):
        if self.gc_disable:
            gc.disable()
        self.starttime = timeit.default_timer()

    def stop(self):
        self.stoptime = timeit.default_timer()
        gc.enable()

    def seconds(self):
        return self.stoptime - self.starttime


def get_filter_read_schema():
    fields = [pa.field("First", pa.string(), False),
              pa.field("Last", pa.string(), False),
              pa.field("Zip", pa.uint32(), False)]

    return pa.schema(fields)


def get_filter_write_schema():
    field = pa.field("First", pa.string(), False)
    return pa.schema([field], metadata={"fletcher_mode": "write"})


def create_batch_from_frame_basic(frame):
    """Converts Pandas dataframe to Arrow batch using the basic from_arrays() function
    Args:
        frame: dataframe to convert

    Returns:
        Arrow RecordBatch

    """
    schema = get_filter_read_schema()

    first_names = pa.array(frame["First"])
    last_names = pa.array(frame["Last"])
    zip_codes = pa.array(frame["Zip"])

    return pa.RecordBatch.from_arrays([first_names, last_names, zip_codes], schema)


def create_batch_from_frame_fast(frame, nthreads=None):
    """Converts Pandas dataframe to Arrow batch in parallel.

    Args:
        frame: dataframe to convert
        nthreads: Number of CPU threads to use

    Returns:
        Arrow RecordBatch

    """
    fields = [pa.field("First", pa.string(), False),
              pa.field("Last", pa.string(), False),
              pa.field("Zip", pa.uint32(), False)]

    schema = pa.schema(fields)

    return pa.RecordBatch.from_pandas(df=frame, schema=schema, preserve_index=False, nthreads=nthreads)


def filter_dataframe_python(frame, zip_code):
    """Filter a dataframe using standard Pandas syntax.

    The filter removes rows that contain last_name="Smith" and zip_code as selected.

    Args:
        frame: Frame to be filtered
        zip_code: Special zip_code to filter on.

    Returns:
        Pandas dataframe containing only the column "First". All rows containing last_name="Smith" & zip_code=zip_code
        are filtered out.

    """
    return frame.loc[(frame["Last"] == "Smith") & (frame["Zip"] == zip_code), ["First"]]


def filter_record_batch_python(batch, zip_code):
    """Filter a RecordBatch using a rather hacky Python loop.

    The filter removes rows that contain last_name="Smith" and zip_code as selected.

    Args:
        batch: bach to be filtered
        zip_code: Special zip_code to filter on.

    Returns:
        Arrow RecordBatch containing only the column "First". All rows containing last_name="Smith" & zip_code=zip_code
        are filtered out.

    """
    first_names = []

    for i in range(batch.num_rows):
        if batch.column(1)[i] == "Smith" and batch.column(2)[i] == zip_code:
            first_names.append(batch.column(0)[i].as_py())

    schema = get_filter_write_schema()

    return pa.RecordBatch.from_arrays([pa.array(first_names, type=pa.string())], schema)


def get_filter_output_rb(offsets_buffer_size, values_buffer_size):
    string_array_length = offsets_buffer_size / 4 - 1
    offsets_buffer = pa.allocate_buffer(offsets_buffer_size)
    values_buffer = pa.allocate_buffer(values_buffer_size)

    array = pa.StringArray.from_buffers(string_array_length, offsets_buffer, values_buffer)

    return pa.RecordBatch.from_arrays([array], get_filter_write_schema())


def filter_record_batch_fpga(batch_in, zip_code, platform_type, offsets_buffer_out_size, values_buffer_out_size, t_copy,
                             t_fpga):
    t = Timer()
    platform = pf.Platform(platform_type)
    context = pf.Context(platform)
    uc = pf.UserCore(context)

    batch_out = get_filter_output_rb(offsets_buffer_out_size, values_buffer_out_size)

    platform.init()

    context.queue_record_batch(batch_in)
    context.queue_record_batch(batch_out)

    uc.reset()
    t.start()
    context.enable()
    t.stop()
    t_copy.append(t.seconds())
    uc.set_range(0, batch_in.num_rows)
    uc.set_arguments([zip_code])

    t.start()
    uc.start()
    uc.wait_for_finish(10)
    t.stop()
    t_fpga.append(t.seconds())

    platform.copy_device_to_host(context.get_buffer_device_address(3, 0), offsets_buffer_out_size,
                           batch_out.column(0).buffers()[1])
    platform.copy_device_to_host(context.get_buffer_device_address(3, 1), values_buffer_out_size,
                                 batch_out.column(0).buffers()[2])

    return batch_out


if __name__ == "__main__":
    t = Timer(gc_disable=False)

    parser = argparse.ArgumentParser()
    parser.add_argument("--num_exp", dest="ne", default=1,
                        help="Number of experiments to perform")
    parser.add_argument("--input_strings", dest="input_file", default="../cmake-build-debug/rows1024.dat",
                        help="Path to file with strings to search through")
    parser.add_argument("--platform_type", dest="platform_type", default="echo", choices=["echo", "aws"],
                        help="Type of FPGA platform")
    parser.add_argument("--ser_threads", dest="ser_threads", default=1,
                        help="Amount of threads to use for Pandas to Arrow serialization")
    parser.add_argument("--special_zip_code", dest="special_zip_code", default=1337,
                        help="Special zip code to filter on.")
    args = parser.parse_args()

    # Parsed args
    ne = int(args.ne)
    input_file = args.input_file
    platform_type = args.platform_type
    ser_threads = int(args.ser_threads)
    special_zip_code = int(args.special_zip_code)

    frame = pd.read_csv(input_file, header=None, dtype={0: np.str, 1: np.str, 2: np.uint32}, keep_default_na=False)
    frame.columns = ["First", "Last", "Zip"]

    # Timers
    t_ser_basic = []
    t_ser_fast = []
    t_pd_py = []
    t_pd_cy = []
    t_pa_py = []
    t_pa_cpp = []
    t_fpga = []
    t_copy = []
    t_ftot = []

    # Results
    r_pd_py = []
    r_pd_cy = []
    r_pa_py = []
    r_pa_cpp = []
    r_fpga = []

    for i in range(ne):
        t.start()
        batch_basic = create_batch_from_frame_basic(frame)
        t.stop()
        t_ser_basic.append(t.seconds())

        t.start()
        batch_fast = create_batch_from_frame_fast(frame, ser_threads)
        t.stop()
        t_ser_fast.append(t.seconds())

    print("Total Pandas to Arrow serialization times for " + str(ne) + " runs: ")
    print("Pandas to Arrow using from_arrays(): " + str(sum(t_ser_basic)))
    print("Pandas to Arrow using from_pandas(): " + str(sum(t_ser_fast)))

    for i in range(ne):
        # Normal Pandas filtering. Creates copy.
        t.start()
        r_pd_py.append(filter_dataframe_python(frame, special_zip_code))
        t.stop()
        t_pd_py.append(t.seconds())

        # Normal Pandas filtering. Creates copy.
        t.start()
        r_pd_cy.append(filter_custom.filter_dataframe_python(frame, special_zip_code))
        t.stop()
        t_pd_cy.append(t.seconds())

        # Filter an Arrow record batch using only pure Python. Creates copy.
        t.start()
        r_pa_py.append(filter_record_batch_python(batch_basic, special_zip_code))
        t.stop()
        t_pa_py.append(t.seconds())

        # Filter an Arrow record batch using Cython wrapped CPP. Creates copy.
        t.start()
        r_pa_cpp.append(filter_custom.filter_record_batch_cpp(batch_basic, special_zip_code))
        t.stop()
        t_pa_cpp.append(t.seconds())

        # Filter an Arrow record batch using the FPGA. Creates copy.
        # Currently requires prior knowledge on the size of the output batch.
        t.start()
        r_fpga.append(filter_record_batch_fpga(batch_basic, special_zip_code, platform_type,
                                               r_pa_cpp[0].column(0).buffers()[1].size,
                                               r_pa_cpp[0].column(0).buffers()[2].size,
                                               t_copy,
                                               t_fpga))
        t.stop()
        t_ftot.append(t.seconds())

        print("Total execution times for " + str(ne) + " runs:")
        print("Pandas pure Python filtering: " + str(sum(t_pd_py)))
        print("Pandas Cython (?) filtering: " + str(sum(t_pd_cy)))
        print("Arrow pure Python filtering: " + str(sum(t_pa_py)))
        print("Arrow CPP filtering: " + str(sum(t_pa_cpp)))
        print("Filter Arrow FPGA copy time: " + str(sum(t_copy)))
        print("Filter Arrow FPGA algorithm time: " + str(sum(t_fpga)))
        print("Filter Arrow FPGA total time: " + str(sum(t_ftot)))
        print()
        print("Average execution times:")
        print("Pandas pure Python filtering: " + str(sum(t_pd_py) / (i + 1)))
        print("Pandas Cython (?) filtering: " + str(sum(t_pd_cy) / (i + 1)))
        print("Arrow pure Python filtering: " + str(sum(t_pa_py) / (i + 1)))
        print("Arrow CPP filtering: " + str(sum(t_pa_cpp) / (i + 1)))
        print("Filter Arrow FPGA copy time: " + str(sum(t_copy)/(i+1)))
        print("Filter Arrow FPGA algorithm time: " + str(sum(t_fpga)/(i+1)))
        print("Filter Arrow FPGA total time: " + str(sum(t_ftot)/(i+1)))

    with open("Output.txt", "w") as textfile:
        textfile.write("\nTotal execution times for " + str(ne) + " runs:")
        textfile.write("\nPandas pure Python filtering: " + str(sum(t_pd_py)))
        textfile.write("\nPandas Cython (?) filtering: " + str(sum(t_pd_cy)))
        textfile.write("\nArrow pure Python filtering: " + str(sum(t_pa_py)))
        textfile.write("\nArrow CPP filtering: " + str(sum(t_pa_cpp)))
        textfile.write("\nFilter Arrow FPGA copy time: " + str(sum(t_copy)))
        textfile.write("\nFilter Arrow FPGA algorithm time: " + str(sum(t_fpga)))
        textfile.write("\nFilter Arrow FPGA total time: " + str(sum(t_ftot)))
        textfile.write("\n")
        textfile.write("\nAverage execution times:")
        textfile.write("\nPandas pure Python filtering: " + str(sum(t_pd_py) / ne))
        textfile.write("\nPandas Cython (?) filtering: " + str(sum(t_pd_cy) / ne))
        textfile.write("\nArrow pure Python filtering: " + str(sum(t_pa_py) / ne))
        textfile.write("\nArrow CPP filtering: " + str(sum(t_pa_cpp) / ne))
        textfile.write("\nFilter Arrow FPGA copy time: " + str(sum(t_copy)/(i+1)))
        textfile.write("\nFilter Arrow FPGA algorithm time: " + str(sum(t_fpga)/(i+1)))
        textfile.write("\nFilter Arrow FPGA total time: " + str(sum(t_ftot)/(i+1)))

    # Check if results are equal.
    pass_counter = 0
    cross_exp_pass_counter = 0
    for i in range(ne):
        if r_pa_py[i].equals(pa.RecordBatch.from_pandas(r_pd_py[i], get_filter_write_schema(), preserve_index=False)) \
                and r_pa_py[i].equals(r_pa_cpp[i]) \
                and r_pa_py[i].equals(pa.RecordBatch.from_pandas(r_pd_cy[i], get_filter_write_schema(), preserve_index=False)) \
                and r_pa_py[i].equals(r_fpga[i]):
            pass_counter += 1

        if r_pa_py[0].equals(r_pa_py[i]):
            cross_exp_pass_counter += 1

    if pass_counter == ne and cross_exp_pass_counter == ne:
        print("PASS")
    else:
        print("ERROR ({error_counter} errors)".format(error_counter=ne - pass_counter))
